packages/truffle-db/src/loaders/artifacts/index.ts

```diff
@@ -0,0 +1,255 @@
+    sources: [CompilationSourceInput!]!
+}  
+input CompilationsAddInput {
+  compilations: [CompilationInput!]!
+}
+
+
+mutation AddCompilation($compilations: [CompilationInput!]!) {
+  workspace {
+    compilationsAdd(input: {
+      compilations: $compilations
+    }) {
+      compilations {
+        id
+        compiler {
+          name
+          version
+        }
+        contracts {
+          name
+          source {
+            contents
+            sourcePath
+          }
+          ast {
+            json
+          }
+        }
+        sources {
+          contents
+          sourcePath
+        }
+      }
+    }
+  }
+}`
+
+export class ArtifactsLoader {
+  private db: TruffleDB;
+  private config: object;
+
+  constructor (db: TruffleDB, config?: object) {
+    this.db = db;
+    this.config = config;
+  }
+
+  async load (): Promise<void> {
+    const {
+      data: {
+        artifacts: {
+          contractNames
+        }
+      }
+    } = await this.db.query(GetContractNames);
+
+    if(this.config) {
+      await this.loadCompilations(this.config);
+    }
+    await this.loadBytecodes(contractNames);
+    await this.loadSources(contractNames);
+
+  }
+
+  async loadBytecodes(contractNames: string[]) {
+    const createBytecodes = await Promise.all(contractNames.map(
+      async (name) =>
+        (await this.db.query(GetBytecode, { name }))
+          .data
+          .artifacts
+          .contract
+          .constructor
+          .createBytecode
+    ));
+
+    const bytecodes = [...createBytecodes];
+
+    await this.db.query(AddBytecodes, { bytecodes });
+  }
+
+  async loadSources(contractNames: string[]) {
+    const contractSources = await Promise.all(contractNames.map(
+      async (name) =>
+        (await this.db.query(GetSource, { name }))
+          .data
+          .artifacts
+          .contract
+          .sourceContract
+          .source
+    ));
+
+    const sources = [...contractSources];
+
+    await this.db.query(AddSources, { sources });
+  }
+  
+  async loadCompilations(compilationConfig: object) {
+    let compilationsArray = [];
+    let sources = [];
+    await Contracts.compile(compilationConfig, async (err, output) => {
+      if(err) console.error(err);
+      else {
+        const compilationsData = Object.values(output.contracts);
+        for(let contract in compilationsData) {
+          //add source from compilation into workspace first, otherwise compilation addition will fail
+          let sourceObject = {
+            contents: compilationsData[contract]["source"], 
+            sourcePath: compilationsData[contract]["sourcePath"]
+          };
+
+          let sourceId = generateId({
+            contents: compilationsData[contract]["source"], 
+            sourcePath: compilationsData[contract]["sourcePath"]
+          });
+          
+          sources.push({contents: compilationsData[contract]["source"], sourcePath: compilationsData[contract]["sourcePath"]});
```

Comment:
Thinking we should move this source stuff to a separate `await this.loadSources()` method, called by `loadCompilations`. (just to clean up the code)